### PR TITLE
refactor: Pass default user context for APIs in the framework middleware itself

### PR DESCRIPTION
**Pull request by KShivendu**
## Summary of change

Pass default user context for APIs in the framework middleware itself

## Related Issues

- https://github.com/supertokens/supertokens-node/pull/595

## Checklist for important updates

-   [ ] Changelog has been updated
-   [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `supertokens_python/constants.py`
-   [ ] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `setup.py`
    -   In `supertokens_python/constants.py`
-   [x] Had installed and ran the pre-commit hook
-   [x] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `supertokens_python/utils.py` file to include that in the `FRAMEWORKS` variable
-   [ ] If added a new recipe that has a User type with extra info, then be sure to change the User type in supertokens_python/types.py
-   [ ] Make sure that `syncio` / `asyncio` functions are consistent.

### CODE REVIEW

File: supertokens_python/recipe/dashboard/recipe.py [hunk 1]
```diff
@@ -133,6 +133,7 @@ async def handle_api_request(
         path: NormalisedURLPath,
         method: str,
         response: BaseResponse,
+        user_context: Dict[str, Any],
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/dashboard/recipe.py:136]: you need to pass in the user_context to the functions in this API. See node PR

File: supertokens_python/recipe/dashboard/recipe.py [hunk 2]
```diff
@@ -247,7 +248,7 @@ def reset():
         DashboardRecipe.__instance = None
 
     def return_api_id_if_can_handle_request(
-        self, path: NormalisedURLPath, method: str
+        self, path: NormalisedURLPath, method: str, user_context: Dict[str, Any]
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/dashboard/recipe.py:251]: missing changes for this function - see node PR

File: supertokens_python/recipe/emailpassword/recipe.py [hunk 3]
```diff
@@ -175,6 +175,7 @@ async def handle_api_request(
         path: NormalisedURLPath,
         method: str,
         response: BaseResponse,
+        user_context: Dict[str, Any],
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/emailpassword/recipe.py:178]: you need to pass in this to the apis called in this.. see node pr

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens_python/recipe/dashboard/recipe.py (hunk 1 above)
Comment (KShivendu) [reply supertokens_python/recipe/dashboard/recipe.py:137]: Fixed.

File: supertokens_python/recipe/dashboard/recipe.py (hunk 2 above)
Comment (KShivendu) [reply supertokens_python/recipe/dashboard/recipe.py:256]: Fixed.

File: supertokens_python/recipe/emailpassword/recipe.py (hunk 3 above)
Comment (KShivendu) [reply supertokens_python/recipe/emailpassword/recipe.py:178]: Fixed.

File: supertokens_python/recipe/multitenancy/interfaces.py [hunk 4]
```diff
@@ -158,8 +158,8 @@ def __init__(self):
 
     @abstractmethod
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/multitenancy/interfaces.py:159]: unrelated to this PR, but i found this type issue -> in create_or_update_tenant in multitenancy syncio and asyncio files, we should make `config` Optional as well (in node sdk, it's optional too).

File: supertokens_python/recipe/session/framework/django/asyncio/__init__.py [hunk 5]
```diff
@@ -37,18 +38,21 @@ def verify_session(
     ] = None,
     user_context: Union[None, Dict[str, Any]] = None,
 ) -> Callable[[_T], _T]:
-    if user_context is None:
-        user_context = {}
+    _ = user_context
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/session/framework/django/asyncio/__init__.py:41]: why is this needed?

File: supertokens_python/recipe/session/framework/django/asyncio/__init__.py [hunk 6]
```diff
@@ -37,18 +38,21 @@ def verify_session(
...
 
         @wraps(f)
         async def wrapped_function(request: HttpRequest, *args: Any, **kwargs: Any):
+            nonlocal user_context
             from django.http import JsonResponse
 
             try:
                 baseRequest = DjangoRequest(request)
+                if user_context is None:
+                    user_context = default_user_context(baseRequest)
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/session/framework/django/asyncio/__init__.py:54]: this is wrong. We want to modify the user_context even if it's passed by the user to add the default stuff into their passed in user_context.

File: supertokens_python/recipe/session/framework/django/syncio/__init__.py [hunk 7]
```diff
@@ -38,18 +39,21 @@ def verify_session(
     ] = None,
     user_context: Union[None, Dict[str, Any]] = None,
 ) -> Callable[[_T], _T]:
-    if user_context is None:
-        user_context = {}
+    _ = user_context
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/session/framework/django/syncio/__init__.py:42]: why is this needed?

File: supertokens_python/recipe/session/framework/django/syncio/__init__.py [hunk 8]
```diff
@@ -38,18 +39,21 @@ def verify_session(
...
 
         @wraps(f)
         def wrapped_function(request: HttpRequest, *args: Any, **kwargs: Any):
+            nonlocal user_context
             from django.http import JsonResponse
 
             try:
                 baseRequest = DjangoRequest(request)
+                if user_context is None:
+                    user_context = default_user_context(baseRequest)
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/session/framework/django/syncio/__init__.py:55]: this is wrong. We want to modify the user_context even if it's passed by the user to add the default stuff into their passed in user_context.

File: supertokens_python/recipe/session/framework/fastapi/__init__.py [hunk 9]
```diff
@@ -32,12 +33,15 @@ def verify_session(
     ] = None,
     user_context: Union[None, Dict[str, Any]] = None,
 ) -> Callable[..., Coroutine[Any, Any, Union[SessionContainer, None]]]:
-    if user_context is None:
-        user_context = {}
+    _ = user_context
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/session/framework/fastapi/__init__.py:36]: why is this needed?

File: supertokens_python/recipe/session/framework/fastapi/__init__.py [hunk 10]
```diff
@@ -32,12 +33,15 @@ def verify_session(
...
     from fastapi import Request
 
     async def func(request: Request) -> Union[SessionContainer, None]:
+        nonlocal user_context
         baseRequest = FastApiRequest(request)
+        if user_context is None:
+            user_context = default_user_context(baseRequest)
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/session/framework/fastapi/__init__.py:43]: this is wrong. We want to modify the user_context even if it's passed by the user to add the default stuff into their passed in user_context.

File: supertokens_python/recipe/session/framework/flask/__init__.py [hunk 11]
```diff
@@ -35,15 +36,18 @@ def verify_session(
     ] = None,
     user_context: Union[None, Dict[str, Any]] = None,
 ) -> Callable[[_T], _T]:
-    if user_context is None:
-        user_context = {}
+    _ = user_context
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/session/framework/flask/__init__.py:39]: why is this needed?

File: supertokens_python/recipe/session/framework/flask/__init__.py [hunk 12]
```diff
@@ -35,15 +36,18 @@ def verify_session(
...
 
     def session_verify(f: _T) -> _T:
         @wraps(f)
         def wrapped_function(*args: Any, **kwargs: Any):
+            nonlocal user_context
             from flask import make_response, request
 
             baseRequest = FlaskRequest(request)
+            if user_context is None:
+                user_context = default_user_context(baseRequest)
```
Comment (rishabhpoddar) [inline supertokens_python/recipe/session/framework/flask/__init__.py:49]: this is wrong. We want to modify the user_context even if it's passed by the user to add the default stuff into their passed in user_context.

File: supertokens_python/supertokens.py [hunk 13]
```diff
@@ -467,7 +467,7 @@ async def update_or_delete_user_id_mapping_info(  # pylint: disable=no-self-use
         raise_general_exception("Please upgrade the SuperTokens core to >= 3.15.0")
 
     async def middleware(  # pylint: disable=no-self-use
-        self, request: BaseRequest, response: BaseResponse
+        self, request: BaseRequest, response: BaseResponse, user_context: Dict[str, Any]
     ) -> Union[BaseResponse, None]:
         log_debug_message("middleware: Started")
         path = Supertokens.get_instance().app_info.api_gateway_path.append(
```
Comment (rishabhpoddar) [inline supertokens_python/supertokens.py:473]: add missing test.

Comment (rishabhpoddar) [reply supertokens_python/supertokens.py:473]: should be 7 occurances

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: supertokens_python/recipe/session/framework/django/asyncio/__init__.py (hunk 5 above)
Comment (KShivendu) [reply supertokens_python/recipe/session/framework/django/asyncio/__init__.py:41]: Usage of `user_context` is 2nd order nested and somehow linter's feel it's not used. 

On removing this line we get:
```
supertokens_python/recipe/session/framework/django/asyncio/__init__.py:39:4: W0613: Unused argument 'user_context' (unused-argument)
```

File: supertokens_python/recipe/session/framework/django/asyncio/__init__.py (hunk 6 above)
Comment (KShivendu) [reply supertokens_python/recipe/session/framework/django/asyncio/__init__.py:54]: Okay. Updated it. Howevere, nodejs also does `const userContext = makeDefaultUserContextFromAPI(request);` which isn't taking userContext anywhere.

File: supertokens_python/recipe/session/framework/django/syncio/__init__.py (hunk 7 above)
Comment (KShivendu) [reply supertokens_python/recipe/session/framework/django/syncio/__init__.py:42]: Usage of user_context is 2nd order nested and somehow linter say it's not used. So they throw an error.

File: supertokens_python/recipe/session/framework/fastapi/__init__.py (hunk 9 above)
Comment (KShivendu) [reply supertokens_python/recipe/session/framework/fastapi/__init__.py:36]: Usage of user_context is 2nd order nested and somehow linter say it's not used. So they throw an error.

File: supertokens_python/recipe/session/framework/flask/__init__.py (hunk 11 above)
Comment (KShivendu) [reply supertokens_python/recipe/session/framework/flask/__init__.py:39]: Usage of user_context is 2nd order nested and somehow linter say it's not used. So they throw an error.

File: supertokens_python/recipe/session/framework/fastapi/__init__.py (hunk 10 above)
Comment (KShivendu) [reply supertokens_python/recipe/session/framework/fastapi/__init__.py:43]: Updated.

File: supertokens_python/recipe/session/framework/django/syncio/__init__.py (hunk 8 above)
Comment (KShivendu) [reply supertokens_python/recipe/session/framework/django/syncio/__init__.py:55]: Updated.

File: supertokens_python/recipe/session/framework/flask/__init__.py (hunk 12 above)
Comment (KShivendu) [reply supertokens_python/recipe/session/framework/flask/__init__.py:49]: Updated

File: supertokens_python/recipe/multitenancy/interfaces.py (hunk 4 above)
Comment (KShivendu) [reply supertokens_python/recipe/multitenancy/interfaces.py:159]: Created https://github.com/supertokens/supertokens-python/pull/377 to fix this.

File: supertokens_python/supertokens.py (hunk 13 above)
Comment (KShivendu) [reply supertokens_python/supertokens.py:473]: Done

File: supertokens_python/recipe/session/framework/django/asyncio/__init__.py (hunk 6 above)
Comment (rishabhpoddar) [reply supertokens_python/recipe/session/framework/django/asyncio/__init__.py]: Thats cause in node, user can't pass userContext into verifySession at all

**Review by rishabhpoddar — APPROVED** [review]
